models/migrations/v58.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/go-xorm/xorm"
+)
+
+func mysqlColumnsToUTF8MB4(x *xorm.Engine) (err error) {
+	switch {
+	case setting.UseMySQL:
+		tables, err := x.DBMetas()
+		if err != nil {
+			return fmt.Errorf("cannot get tables: %v", err)
+		}
+		for _, table := range tables {
+			for ix := range table.Indexes {
+				if col := table.GetColumn(ix); col != nil {
+					if col.SQLType.Name == "VARCHAR" && col.Length > 191 {
+						log.Info("%s.%s: reducing field length from %d to 191", table.Name, ix, col.Length)
+						if _, err := x.Exec("alter table `" + table.Name + "` change column `" + ix + "` `" + ix + "` varchar(191)"); err != nil {
+							return fmt.Errorf("failed to reduce column %s.%s length: %v", table, ix, err)
+						}
+					}
+				}
+			}
+			log.Info("%s: converting table to utf8mb4", table.Name)
+			if _, err := x.Exec("alter table `" + table.Name + "` convert to character set utf8mb4"); err != nil {
+				return fmt.Errorf("conversation of %s failed: %v", table, err)
+			}
+		}
+	default:
+		log.Info("Nothing to do")
```

Comment:
at the top 

```go
if !setting.UseMySQL {
log.Info("Nothing to do")
return nil
```

and remove switch?

Reply:
great idea! I'll provide the requested changes tomorrow.